File: problems/592_fraction_addition_and_subtraction/Solution.java
class Solution {
    public String fractionAddition(String expression) {
        long num = 0, den = 1;
        int i = 0, n = expression.length();
        while (i < n) {
            int sign = 1;
            if (expression.charAt(i) == '+' || expression.charAt(i) == '-') {
                if (expression.charAt(i) == '-') sign = -1;
                i++;
            }
            long numer = 0;
            while (i < n && Character.isDigit(expression.charAt(i)))
                numer = numer * 10 + (expression.charAt(i++) - '0');
            numer *= sign;
            i++; // skip '/'
            long denom = 0;
            while (i < n && Character.isDigit(expression.charAt(i)))
                denom = denom * 10 + (expression.charAt(i++) - '0');
            num = num * denom + numer * den;
            den *= denom;
            long g = gcd(Math.abs(num), Math.abs(den));
            num /= g; den /= g;
        }
        if (den < 0) { num = -num; den = -den; }
        return num + "/" + den;
    }
    private long gcd(long a, long b) { return b == 0 ? a : gcd(b, a % b); }
}

File: problems/592_fraction_addition_and_subtraction/solution.py
class Solution:
    def fractionAddition(self, expression: str) -> str:
        import re
        from math import gcd
        tokens = re.findall(r'[+-]?\d+/\d+', expression)
        num, den = 0, 1
        for token in tokens:
            n, d = map(int, token.split('/'))
            num = num * d + n * den
            den *= d
            g = gcd(abs(num), abs(den))
            num //= g
            den //= g
        if den < 0:
            num, den = -num, -den
        return f'{num}/{den}'
